src/bsoncxx/test/optional.test.cpp

```diff
@@ -0,0 +1,250 @@
+    return check_conversions<T, T>();
+}
+
+}  // namespace
+
+static_assert(bsoncxx::detail::is_totally_ordered<std::string>{}, "fail");
+static_assert(bsoncxx::detail::is_totally_ordered<int>{}, "fail");
+static_assert(!bsoncxx::detail::is_totally_ordered<not_ordered>{}, "fail");
+
+TEST_CASE("Trait checks") {
+    CHECK(static_checks<int>());
+    CHECK(static_checks<std::string>());
+    CHECK(static_checks<std::unique_ptr<int>>());
+    CHECK(static_checks<immobile>());
+    CHECK(static_checks<not_copyable>());
+    CHECK(static_checks<not_default_constructible>());
+    CHECK(static_checks<allows_moving_explicit_conversion>());
+    CHECK(static_checks<allows_moving_implicit_conversion>());
+    CHECK(check_conversions<int, int>());
+    CHECK(check_conversions<int, double>());
+    CHECK(check_conversions<double, std::string>());
+    CHECK(check_conversions<std::string, double>());
+    CHECK(check_conversions<std::string, const char*>());
+    CHECK(check_conversions<const char*, std::string>());
+    CHECK(check_conversions<int*, std::unique_ptr<int>>());
+    CHECK(check_conversions<int, allows_moving_explicit_conversion>());
+    CHECK(check_conversions<int, allows_moving_implicit_conversion>());
+    optional<int> v(21);
+    CHECK(*v == 21);
+    CHECK_FALSE(v < v);
+}
+
+TEST_CASE("optional: Nontrivial contents") {
+    optional<std::string> str = "abcd1234";
+    CHECK(str == "abcd1234");
+    {
+        auto dup = str;
+        bool b = dup == str;
+        (void)b;
+        CHECK(dup == str);
+    }
+    CHECK(str == "abcd1234");
+
+    optional<std::unique_ptr<int>> aptr;
+    CHECK(aptr != nullptr);
+    CHECK(aptr == bsoncxx::stdx::nullopt);
+    {
+        auto dup = std::move(aptr);
+        CHECK(aptr == dup);
+    }
+    aptr = bsoncxx::stdx::make_unique<int>(31);
+    CHECK(aptr != nullopt);
+    REQUIRE(aptr != nullptr);
+    CHECK(**aptr == 31);
+    {
+        auto dup = std::move(aptr);
+        CHECK(aptr);
+        CHECK(aptr == nullptr);
+        REQUIRE(dup);
+        REQUIRE(dup != nullptr);
+        CHECK(**dup == 31);
+    }
+    CHECK(aptr == nullptr);
+}
+
+TEST_CASE("Comparisons") {
+    optional<int> a = 21;
+    optional<int> b = 23;
+    optional<int> c;
+    CHECK(a != nullopt);
+    CHECK(a == 21);
+    CHECK(b != nullopt);
+    CHECK(b == 23);
+    CHECK(c == nullopt);
+    // Null compares less-than values:
+    CHECK(c < 42);
+    CHECK(a < b);
+    CHECK(c < a);
+    CHECK(c < b);
+    CHECK(a != b);
+    CHECK(c != a);
+    CHECK(c != b);
+    CHECK(c == c);
+
+    optional<double> dbl = 3.14;
+    CHECK(dbl != a);
+}
+
+struct in_place_convertible {
+    bool constructed_from_in_place = false;
+    in_place_convertible() = default;
+    in_place_convertible(bsoncxx::stdx::in_place_t) : constructed_from_in_place(true) {}
+};
+
+TEST_CASE("optional<T> conversions") {
+    static_assert(!std::is_constructible<optional<bool>, optional<std::string>>{}, "fail");
```

Comment:
Implementations disagree. MSVC STL rejects the conversion, while libstdc++ accepts it. I'm inclined to believe that Microsoft is correct in this case, as the intention appears to be to prevent optional's implicit-bool from jumping in to satisfy the converting constructor.

Reply:
Interesting. Suggest leaving a comment to describe the motivation of this test:

```c++
    // Expect failure to convert `optional<bool>` to `optional<std::string>`.
    // Note: standard library implementations disagree on this behavior.
    // MSVC STL rejects the conversion, while libstdc++ accepts it.
    static_assert(!std::is_constructible<optional<bool>, optional<std::string>>{}, "fail");
```

Reply:
I also believe MSVC is correct.

There is [a long list](https://timsong-cpp.github.io/cppwp/n4659/optional.ctor#31) of constraints that `T` and `U` must satisfy for the "from optional<U>" constructor to participate in overload resolution. Given `T = bool` and `U = std::string`, the very first constraint `is_constructible_v<T, U&&>` is false (`bool` is not constructible from a `std::string&&`), therefore this ctor should not participate in overload resolution.

Likewise, the "from U" constructor also should not participate in overload resolution due to the `is_constructible_v<T, U&&>` constraint being false (same as above).

There are no other possible ctor candidates that can accept a single `optional<U>` argument. Therefore, the original assertion is correct that an `optional<bool>` should not be constructible from an `optional<std::string>`.

I prefer keeping this assertion for our implementation even if it may conflict with some `std::` implementations.